File: c/08_BSTree(demo).c
/*
    이진 탐색 트리(BST): 알고리즘 구현
        파일명: BSTree(demo).cpp
            - main: 삽입(Insert), 삭제(Delete), 검색(Search), 전체 원소 출력
*/

#include <stdio.h>
#include <stdlib.h>			// exit
#include <stdbool.h>		// bool, true, false
#include "BSTree.h"			// DNode
#define	 bufferMAXSIZE 1024

DNode	*InsertBST(DNode *root);
void	SearchBST(DNode *root);
void	DeleteBST(DNode *root);
void	printBSTAll(DNode* root);

int main(void)
{
	int		num;
	DNode	*root = NULL;
	while (true) {
		system("cls");
		printf("\n ### 이진 검색 트리 ### \n\n");
		printf("1) 데이터 삽입 \n");
		printf("2) 데이터 검색 \n");
		printf("3) 데이터 삭제 \n");
		printf("4) 전체 원소 출력 \n");
		printf("5) 프로그램 종료 \n");
		printf("\n메뉴 선택: ");
		scanf_s("%d%*c", &num);
		// scanf("%d%*c", &num);

		switch (num) {
			case 1: root = InsertBST(root);	break;
			case 2: SearchBST(root);		break;
			case 3: DeleteBST(root);		break;
			case 4: printBSTAll(root);		break;
			case 5: printf("프로그램 종료!!!\n");
					return 0;
			default: printf("메뉴를 잘못 선택하셨습니다. \n");
		}
		system("pause");
	}
	return 0;
}

// 이진 검색 트리(BST): 데이터 입력
DNode	*InsertBST(DNode *root) {
	int	num;
	while (true) {
		printf("임의의 정수 입력 (종료 : 0) : ");
		scanf_s("%d%*c", &num);
		// scanf("%d%*c", &num);
		if (num == 0)
			break;

		// 데이터 삽입
		root = insertBST(root, num);
	}
	return root;
}

// 이진 검색 트리(BST): 데이터 검색
void  SearchBST(DNode *root) {
	int		num;
	DNode	*tNode = NULL;
	while (true) {
		printf("\n찾을 임의의 정수 값을 입력하세요(종료: 0): ");
		scanf_s("%d%*c", &num);
		// scanf("%d%*c", &num);
		if (num == 0)
			break;

		// 데이터 검색
		tNode = searchBST(root, num);
		if (tNode)	printf("%d 키를 찾았습니다!!! \n", tNode->__data);
		else		printf("키를 찾지 못했습니다. \n");
	}
	return;
}

// 이진 검색 트리(BST): 데이터 삭제
void  DeleteBST(DNode *root) {
	int		num;
	DNode	*tNode = NULL;
	while (true) {
		printf("\n삭제할 임의의 정수 값을 입력하세요(종료: 0): ");
		scanf_s("%d%*c", &num);
		// scanf("%d%*c", &num);
		if (num == 0)
			break;

		// 데이터 삭제
		deleteBST(root, num);
		if (tNode)	printf("%d 키를 찾았습니다!!! \n", tNode->__data);
		else		printf("키를 찾지 못했습니다. \n");
	}
	return;
}

// 이진 검색 트리(BST): 전체 노드 출력(중위 순회)
void printBSTAll(DNode *root) {
	if (root) {
		printBSTAll(root->__Llink);
		printf("%3d", root->__data);
		printBSTAll(root->__Rlink);
	}
}

File: python/08_BSTree(demo).py
class DNode :
    def __init__ (self, data, Llink=None, Rlink=None):
        self.data = data
        self.Llink = Llink
        self.Rlink = Rlink

class BSTree:
	def __init__(self):
		self.__root = None

	# 이진 검색 트리(BST): 데이터 검색
	def search(self, data) -> DNode:
		return self.__searchBST(self.__root, data)
	def __searchBST(self, tNode:DNode, data) -> DNode:

	# 이진 검색 트리(BST): 데이터 삽입
	def insert(self, data):
		self.__root = self.__insertBST(self.__root, data)
	def __insertBST(self, tNode:DNode, data) -> DNode:

	# 이진 탐색 트리(BST): 데이터 삭제
	def delete(self, data):
		self.__root = self.__deleteBST(self.__root, data)	
	def __deleteBST(self, tNode:DNode, data) -> DNode:

	def __deleteNode(self, tNode:DNode) -> DNode:
		# 3가지 case
		#     1. tNode이 리프 노드
		#     2. tNode이 자식이 하나만 있음
		#     3. tNode이 자식이 둘 있음
		if tNode.Llink == None and tNode.Rlink == None:
			return None
		elif tNode.Llink == None:  # case 2(오른쪽 자식 뿐)
			return tNode.Rlink
		elif tNode.Rlink == None: # case 2(왼쪽 자식 뿐)
			return tNode.Llink
		else: # case 3(두 자식이 다 있음)
			(rtnItem, rtnNode) = self.__deleteMinItem(tNode.Rlink)
			tNode.data = rtnItem
			tNode.Rlink = rtnNode
			return tNode

	def __deleteMinItem(self, tNode:DNode) -> tuple:
		if tNode.Llink == None:
			return (tNode.data, tNode.Rlink)
		else:
			(rtnItem, rtnNode) = self.__deleteMinItem(tNode.Llink)
			tNode.Llink = rtnNode
			return (rtnItem, tNode)
		
	# 기타
	def isEmpty(self) -> bool:
		return self.__root == None

	def clear(self):
		self.__root = None

	def printBSTAll(self):
		def _Preorder(tNode):
			if tNode :
				print(tNode.data, end=' ')
				_Preorder(tNode.Llink)
				_Preorder(tNode.Rlink)
		_Preorder(self.__root)

# from BSTree import *
if __name__ == '__main__':
    bst = BSTree()

    bst.insert(10)
    bst.insert(20)
    bst.insert(5)
    bst.insert(80)
    bst.insert(90)
    bst.insert(7550)
    bst.insert(30)
    bst.insert(77)
    bst.insert(15)
    bst.insert(40)
    # bst.printBSTAll()

    bst.delete(7550)
    bst.delete(10)
    # bst.printBSTAll()

